New buildah task for RHTAP

brunoapimentel opened the pull request:
This PR proposes a new simplified and improved buildah task to be used for the new RHTAP docker build  pipeline.

Compared to the current Konflux buildah task, this task:
- Removes all the Konflux-specific features that are not needed in RHTAP
- Uses a productized version of the buildah image
- Removes root access from all steps
- Sets the correct Syft catalogers when scanning the built image
- Pins the output CycloneDX version
- Avoids littering the source code workspace with temporary files
- Improves the SBOM merge algorithm (dedup by PURL)
- Add a the SBOM OCI artifact URL as a task result
- Changes the tags

arewm commented on task/buildah-dance/0.1/buildah-dance.yaml:
```diff
@@ -0,0 +1,253 @@
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.12.1"
+    tekton.dev/tags: "image-build, appstudio, hacbs"
+  name: buildah-dance
```
```suggestion
  name: buildah-rhtap
```

and we can change the directory/filename too.

brunoapimentel replied:
Changed it.

arewm commented on task/buildah-dance/0.1/buildah-dance.yaml:
```diff
@@ -0,0 +1,253 @@
+    build.appstudio.redhat.com/build_type: "docker"
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.12.1"
+    tekton.dev/tags: "image-build, appstudio, hacbs"
```
Do we need to continue pulling the `appstudio` and `hacbs` tags?

brunoapimentel replied:
I'm really not sure what these tags do nowadays. Maybe @mmorhun can help.

chmeliik replied:
Probably nothing? It might be nice to have a common tag for all RHTAP (former Dance) tasks though.

brunoapimentel replied:
I'm still not sure what's the best approach here. Should we remove these and add a new `rhtap` tag? The tekton catalog only has `containers` as [tag](https://github.com/openshift-pipelines/tektoncd-catalog/blob/p/tasks/task-buildah/0.2.0/task-buildah.yaml#L12).

Also, the min version seems off too.

mmorhun replied:
I'd remove outdated stuff. To me only `rhtap` makes sense. We might also need that `containers` for consistency.

brunoapimentel replied:
Ok, I changed the tags to `containers, rhtap`.

arewm commented on task/buildah-dance/0.1/buildah-dance.yaml:
```diff
@@ -0,0 +1,253 @@
+      name: varlibcontainers
+    workingDir: $(workspaces.source.path)
+
+  - name: upload-sbom
```
@lcarva, do you have thoughts on whether we should change the process for uploading SBOMs to the registry for customer utilization? Is `cosign attach sbom` still the best option?

lcarva replied:
Funny thing is that `cosign attach sbom` actually uploads the sbom to the registry as a blob which is great. what we do need is to record the digest of the sbom so we can ensure it has not changed post-build. We had this at some point but had to revert because that was the tipping point for exceeding the size of the results in the buildah task.

In konflux, we can get around this because the sbom is always embedded in the image itself, so EC can just use that.

brunoapimentel replied:
We are removing the embedded SBOM for RHTAP in this PR, so we won't be able to rely on that in case we need to. :grimacing: 
(no changes for Konflux though)

lcarva replied:
Oh... Can we introduce the [SBOM_BLOB_URL](https://github.com/redhat-appstudio/build-definitions/pull/654/files) for this task, then?

I'm happy to do so as a follow up PR.

brunoapimentel replied:
I found this is already a requirement for RHTAP (related [comment](https://github.com/redhat-appstudio/build-definitions/pull/729#discussion_r1445274263)). It is supposed to be renamed to `LINK_TO_SBOM`, but I believe it has the same purpose.

Although, I'm confused on how to implement this now. I entered the [deprecate cosign attachment](https://github.com/sigstore/cosign/issues/2755) rabbit hole once I saw the deprecation message in the CLI, and the task currently still is dependent on attach.

lcarva replied:
My suggestion is to keep using the `cosign attach` command as it is done today. It is deprecated, but right now we don't really have a better alternative. The main thing here is that that command uploads the SBOM to the registry as a "blob". So just record the sh256 digest of the SBOM and we should be able to find it in the registry, e.g. `quay.io/foo/bar@sha256:abc..`.

I can help you with this. Feel free to ping me.

brunoapimentel replied:
So, @lcarva, I took the same [approach](https://github.com/redhat-appstudio/build-definitions/pull/729/commits/9912beaddf1ffd58022a90a3069d20aac66d36a2) to generate the digest as in the PR you linked. Thing is, taking the `sha256sum sbom-cyclonedx.json` will give me the digest of the layer, not of the SBOM image that is pushed to the registry. 

You can see the test I did in https://quay.io/repository/bpimente/test-images:
- the SBOM image digest is `ed244f35bd7959a1149836997cd568361dbae6e496ffa560821e0d29c46c94fc`
- the only layer existing has a digest of `57a8c881a24e554cecd66dce96645b55b331138c9613586c3507482b09f0e3c4` 

I can compute the latter if I
```
cosign download sbom quay.io/bpimente/test-images:test-rhtap > sbom.json
sha256sum sbom.json
```

Maybe I'm doing something wrong?

brunoapimentel replied:
I'll move this to a separate PR so we can wrap up the work on this one here.

lcarva replied:
Bruno and I chatted offline, summarizing it here for posterity/transparency. 

We want to record the digest of the layer (which should match the digest of the SBOM file). This is what EC will fetch. An SBOM image could, in theory, contain multiple SBOMs. Adding an SBOM changes the image digest, but not the digest of the other SBOMs (as layers).

arewm commented on task/buildah-dance/0.1/buildah-dance.yaml:
```diff
@@ -0,0 +1,253 @@
+    securityContext:
+      runAsUser: 0
+
+  - name: inject-sbom-and-push
```
You indicated that you removed this:

> Copying of the SBOM into the built container image

So it it intentional that it is here still or was the comment inaccurate?

brunoapimentel replied:
I did remove the one [line](https://github.com/redhat-appstudio/build-definitions/pull/729/commits/d32f8491341dd368d020c5b47ee324026c9e6b8c#diff-e27d0664a577cb8c73878188508eccd2fe7d26a67e8c4a1687df758f7aae03a6L310) that was embedding the SBOM. The name of the step is misleading, so I changed it.

arewm commented on task/buildah-dance/0.1/buildah-dance.yaml:
```diff
@@ -0,0 +1,253 @@
+    name: IMAGE
+    type: string
+  - default: registry.access.redhat.com/ubi9/buildah@sha256:04fde77ea72c25b56efb3f71db809c5d7b09938130df2da9175a3c888b94043d
+    # per https://kubernetes.io/docs/concepts/containers/images/#imagepullpolicy-defaulting
+    # the cluster will set imagePullPolicy to IfNotPresent
+    # also per direction from Ralph Bean, we want to use image digest based tags to use a cue to automation like dependabot or renovatebot to periodially submit pull requests that update the digest as new images are released.
```
I don't think we need to include these comments in the pipeline definition.

brunoapimentel replied:
Agreed. Removed all of them.

arewm commented on task/buildah-dance/0.1/buildah-dance.yaml:
```diff
@@ -0,0 +1,253 @@
+      name: varlibcontainers
+    workingDir: $(workspaces.source.path)
+
+  - name: upload-sbom
```
Are we properly generating the results/annotations to enable the SBOM to be visualized in the UI? I don't see that present anywhere but maybe I am missing it.

brunoapimentel replied:
What are the exact results is the UI expecting regarding SBOMs?  The buildah task currently emits three results, but they're only related to image digests.

Maybe you were looking for this task [here](https://github.com/redhat-appstudio/build-definitions/blob/main/task/show-sbom/0.1/show-sbom.yaml) here?

arewm replied:
Whatever is needed for https://docs.google.com/document/d/1_1YXFx0ymzjl4b9M_LDjmmGrEYey5mDrfTdNn_56hpM/edit#heading=h.y4al2k1ggzhu

@rohitkrai03, would you expect that here or in `show-sbom`?

brunoapimentel replied:
I was not aware of this at all. Are these annotations supposed to be added to the task itself? And is saving the SBOM URL in a tekton result enough for it to be shown in the UI?

As for the `show-sbom` task, it seems that its purpose is to download the SBOM and show it within the task logs... apparently the design is supposing that both scenarios could exist (link and logs)?

arewm commented on task/buildah-dance/OWNERS:
```diff
@@ -0,0 +1 @@
+Stonesoup Build Team
```
Let's change this too.

brunoapimentel replied:
Done.

chmeliik commented on task/buildah-rhtap/OWNERS:
```diff
@@ -0,0 +1 @@
+RHTAP Build Team
```
I think this should be a github team: https://github.com/orgs/redhat-appstudio/teams

brunoapimentel replied:
Do you think that adding the full url to the github team is enough?

chmeliik replied:
Just the name probably? I thought these were the Kubernetes OWNERS files, but looking at the docs our syntax is not correct so IDK https://www.kubernetes.dev/docs/guide/owners/

brunoapimentel replied:
Well, looking at the other tasks, they just use an internal team name. I don't think we need to set a new standard in this PR. I'll leave a note for this to be improved later.

`Konflux Build Team` seems to be what it is for now, so that's what I picked, but maybe we can stick to `Stonesoup Build Team` and change it the same time the other tasks do. :shrug:

mmorhun replied:
This is out of scope of this PR, let's just have a value for now and align all later.

chmeliik reviewed (commented):
Oh no, taking a good look at our buildah task did not end well. Congratulations @brunoapimentel, you get a review for the whole task :grin: 

Some of the comments are definitely not blockers, we can probably just file new storiesor say it's acceptable

chmeliik commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,237 @@
+  steps:
+  - image: $(params.BUILDER_IMAGE)
+    name: build
+    computeResources:
+      limits:
+        memory: 4Gi
+      requests:
+        memory: 512Mi
+        cpu: 250m
```
Probably not true, clamav-scan also [defines computeResources](https://github.com/redhat-appstudio/build-definitions/blob/54ecd81ce2791a83da05e289bd7630b87e201d55/task/clamav-scan/0.1/clamav-scan.yaml#L43) and I believe users do have a way to override them (right @MartinBasti?)

MartinBasti replied:
Yeah, I cannot override it and steps overrides are only alpha

brunoapimentel replied:
I removed the resources definition from the task.

chmeliik replied:
I don't know if that's the solution - unless we can ensure that every user will set the requests and limits in their own pipelineRun (do the limits get applied if they're set in the pipelineRun)?

If they do get applied, maybe the solution would be:
* document that the user must set requests and limits, suggest defaults
* in the pipelineRun template that we ship, set default requests and limits

brunoapimentel replied:
I'll file a story so we can investigate this further.

arewm commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,237 @@
+    - name: COMMIT_SHA
+      value: $(params.COMMIT_SHA)
+    script: |
+      SOURCE_CODE_DIR=source
+      if [ -e "$SOURCE_CODE_DIR/$CONTEXT/$DOCKERFILE" ]; then
+        dockerfile_path="$SOURCE_CODE_DIR/$CONTEXT/$DOCKERFILE"
+      elif [ -e "$SOURCE_CODE_DIR/$DOCKERFILE" ]; then
+        dockerfile_path="$SOURCE_CODE_DIR/$DOCKERFILE"
+      elif echo "$DOCKERFILE" | grep -q "^https\?://"; then
+        echo "Fetch Dockerfile from $DOCKERFILE"
+        dockerfile_path=$(mktemp --suffix=-Dockerfile)
+        http_code=$(curl -s -L -w "%{http_code}" --output "$dockerfile_path" "$DOCKERFILE")
+        if [ $http_code != 200 ]; then
+          echo "No Dockerfile is fetched. Server responds $http_code"
+          exit 1
+        fi
+        http_code=$(curl -s -L -w "%{http_code}" --output "$dockerfile_path.dockerignore.tmp" "$DOCKERFILE.dockerignore")
+        if [ $http_code = 200 ]; then
+          echo "Fetched .dockerignore from $DOCKERFILE.dockerignore"
+          mv "$dockerfile_path.dockerignore.tmp" $SOURCE_CODE_DIR/$CONTEXT/.dockerignore
+        fi
+      else
+        echo "Cannot find Dockerfile $DOCKERFILE"
+        exit 1
+      fi
```
I don't think it breaks L3, no, but it would probably be better if the references were pinned and we enabled the hash verification of fetched resources. Do we take advantage of this for enabling "s2i" builds for various languages?

brunoapimentel replied:
From the discussions we had, this custom logic was added to enable devfile builds. Since there's no plans to support them in the near future, I dropped it.

chmeliik replied:
Just FYI: EC already blocks Dockerfiles downloaded from urls https://enterprisecontract.dev/docs/ec-policies/release_policy.html#buildah_build_task__buildah_uses_local_dockerfile

lkolacek commented on pipelines/docker-build-dance/patch.yaml:
```diff
@@ -23,7 +23,7 @@
   path: /spec/tasks/3/taskRef
   value:
-    name: buildah
+    name: buildah-dance
```
Should we stick everywhere to rhtap name?

brunoapimentel replied:
Fixed it.

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,237 @@
+    type: string
+  - default: registry.access.redhat.com/ubi9/buildah@sha256:04fde77ea72c25b56efb3f71db809c5d7b09938130df2da9175a3c888b94043d
+    description: The location of the buildah builder image.
+    name: BUILDER_IMAGE
```
I removed this parameter. I don't think we need it to be configurable, and the default buildah task in the catalog does not allow this to be set either.

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,237 @@
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.12.1"
+    tekton.dev/tags: "image-build, appstudio, hacbs"
+  name: buildah-rhtap
```
Sure, if you want the long version, here's my [attempt](https://github.com/redhat-appstudio/build-definitions/commit/1e548f9709015d4a0cd0ea2ffe83b9feadfddbb0) to port the task into build-definitions as it is, and a [version](https://github.com/redhat-appstudio/build-definitions/commit/4466dc6f3fb93fa1d18407ca8a3500c7a6b609af) with the SBOM generation added to it. 

But in summary, the default buidah task:
- validates input params, existence of the Dockerfile and build context
- builds the image
    - can use a dockerconfig secret (optional) 
    - sets the --no-cache option
    - allows for extra args build args to be added
- pushes image 
    - allows for extra push args to be added
    - can be skipped
- emits the results
    - digest sum
    - image name and tag

It is much more straightforward, it doesn't use things such as root access or a mounted `/var/lib/containers` volume, as it does build an push in the same step.

chmeliik replied:
> can use a dockerconfig secret (optional)

I think ours does too, but implicitly (tekton automatically injects some secrets to predefined directories)

> allows for extra args build args to be added

As in `--build-arg`? Or any argument at all? That would give the default buildah task way more flexibility. Proably too much flexibility to be EC-verifiable, since the user could make buildah do just about anything with the extra args

brunoapimentel replied:
Any argument at all... this is what it looks like:
```
buildah \
      --storage-driver="${PARAMS_STORAGE_DRIVER}" \
      --tls-verify="${PARAMS_TLS_VERIFY}" \ 
      bud ${PARAMS_BUILD_EXTRA_ARGS} \
            --no-cache \
            --file="${DOCKERFILE_FULL}" \
            --tag="${PARAMS_IMAGE}" \
            ${PARAMS_CONTEXT}
 ```

chmeliik replied:
Yeah, this would allow them to mount any volumes and secrets and who knows what else. Which we may need to support in the future, but probably in a more controlled way

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,237 @@
+        "--label" "architecture=$(uname -m)"
+        "--label" "vcs-type=git"
+      )
+      [ -n "$COMMIT_SHA" ] && LABELS+=("--label" "vcs-ref=$COMMIT_SHA")
```
For now, I just removed this piece, as it seemed not necessary for our initial implementation.

brunoapimentel commented:
Newer pushes are addressing some of the comments, but there are still quite a few comments to address.

The idea of using the default buildah task from the catalog [proved](https://github.com/redhat-appstudio/build-definitions/compare/main...brunoapimentel:build-definitions:buildah-rhtap?expand=1) not very promising, at least not for the current phase. I think the way the catalog handles tasks (by having a separate [repository](https://github.com/openshift-pipelines/task-containers/) for scripts and templates, and generating the actual tasks with the scripts files added as base64 encoded [strings](https://github.com/openshift-pipelines/tektoncd-catalog/blob/p/tasks/task-buildah/0.2.0/task-buildah.yaml#L130-L132)) is very different from what we currently do, and just adding the raw scripts code proves to be very verbose, not to mention we lose the ability to reuse scripts such as `common.sh`.

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,237 @@
+        json.dump(image_sbom, f, indent=4)
+    workingDir: $(workspaces.source.path)
+    securityContext:
+      runAsUser: 0
```
I don't think so. I changed the steps so they wouldn't need this anymore.

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,237 @@
+  - name: sbom-syft-generate
+    image: quay.io/redhat-appstudio/syft:v0.98.0@sha256:4d3856e6a2622700b9a9d5d74d9aaf5d8a55671653f80bf6c636677658680ede
+    script: |
+      syft dir:$(workspaces.source.path)/source --output cyclonedx-json=$(workspaces.source.path)/sbom-source.json
```
Added.

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,237 @@
+    script: |
+      syft dir:$(workspaces.source.path)/source --output cyclonedx-json=$(workspaces.source.path)/sbom-source.json
+      find $(cat /workspace/container_path) -xtype l -delete
+      syft dir:$(cat /workspace/container_path) --output cyclonedx-json=$(workspaces.source.path)/sbom-image.json
```
Ok, this took me some time to figure out. I think we pontentially have two approaches to solve this:
- Try to use `oci-dir` instead of `dir`. I couldn't figure out a way to get/mount the image as an oci-dir though
- Manually set the catalogers. I added a patch that does that (in a very inelegant way)

brunoapimentel replied:
For the latter approach, since we still use0.98, we can't use tags to set catalogers (introduced in [0.101](https://github.com/anchore/syft/tree/v0.101.0)), such as:
```
syft <some container image> --select-catalogers "image"
```

So in the commit I added, I needed to manually specify the catalogers as describe in the [0.98 docs](https://github.com/anchore/syft/tree/v0.98.0#image-scanning). Dumb me is still struggling with generating a config yaml file from within the script block from within the task yaml file :laughing:

Any hints are appreciated, but for now, I added a very ugly single-liner that solves the problem (apparently, syft also does not like when we break the args in multiple lines using \, I'm still not sure why).

chmeliik replied:
> Try to use oci-dir instead of dir. I couldn't figure out a way to get/mount the image as an oci-dir though

Would this work?

```shell
buildah push "$image" oci-archive:image.tar
syft oci-archive:image.tar
```

brunoapimentel replied:
`oci-archive:` does not do it, but `oci-dir:` works. For reference, these were the catalogers executed this time:
- alpm-db-cataloger
- apk-db-cataloger
- binary-cataloger
- conan-info-cataloger
- dpkg-db-cataloger
- dotnet-portable-executable-cataloger
- go-module-binary-cataloger
- java-archive-cataloger
- graalvm-native-image-cataloger
- javascript-package-cataloger
- nix-store-cataloger
- php-composer-installed-cataloger
- portage-cataloger
- python-installed-package-cataloger
- r-package-cataloger
- rpm-db-cataloger
- ruby-installed-gemspec-cataloger
- sbom-cataloger

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,237 @@
+      base_image_digest=$(buildah inspect --format '{{ index .ImageAnnotations "org.opencontainers.image.base.digest"}}' $IMAGE)
+      container=$(buildah from --pull-never $IMAGE)
+      buildah config -a org.opencontainers.image.base.name=${base_image_name} -a org.opencontainers.image.base.digest=${base_image_digest} $container
+      buildah commit $container $IMAGE
```
Done :feelsgood:

chmeliik commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,237 @@
+
+      buildah build \
+        ${LABELS[@]} \
+        --tls-verify=$TLSVERIFY --no-cache \
```
If the official buildah task does it, we can do it as well. I do believe builds will be isolated even without --pull=always

https://github.com/redhat-appstudio/build-definitions/pull/729#discussion_r1454319886

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,237 @@
+        echo "Pushing sbom image to registry"
+        buildah push \
+          --tls-verify=$TLSVERIFY \
+          --digestfile $(workspaces.source.path)/image-digest $IMAGE \
```
I agree. I changed this so it is only written in the tmpfiles volume.

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,237 @@
+
+      # fetch unique components from available SBOMs
+      def get_identifier(component):
+        return component["name"] + '@' + component.get("version", "")
```
I changed the algorithm so it looks at the `purl` now. Thing is, there is one case in which the purl is always absent: operating-system components. That should be fine, as they're only reported once in the image scan, but it led me to allow any component that lacks a purl to be present in the final SBOM (at risk of duplication).

brunoapimentel replied:
I don't think this inline Python script looks good, though, so I think we should improve Cachi2's merge [script](https://github.com/containerbuildsystem/cachi2/blob/main/utils/merge_syft_sbom.py) to be more generalistic, able to handle any CycloneDX SBOM regardless of the generation tool.

chmeliik replied:
Trying to do general SBOM de-duplication will be rough, because purls => packages is an N:1 relation, what with all the optional qualifiers etc. Some duplicates will always remain.

And if the SBOM generation tool is not careful or doesn't care, purls => packages can also be 1:M. For example, you can't be sure if `pkg:golang/sigs.k8s.io/cluster-api-provider-aws/v2` really means the original module or the [`github.com/openshift/cluster-api-provider-aws`](https://github.com/openshift/cluster-api-provider-aws/blob/4c2d9b8e1626337cdbe8cdd5a0ad43c2d520ecd1/go.mod#L1C8-L1C47) fork (with cachi2 you probably can, with Syft you can't). You also can't be sure if the tool neglected to include a `?repository` qualifier for a package that comes from a non-default repository, `?vcs_url` for a package that comes straight from git, etc. In this case, by trusting the purl you end up discarding non-duplicate packages, which may or may not be fine.

brunoapimentel replied:
Yeah, excelent point. So I think we'll need to keep doing a tool-based dedup, since we can better predict its behavior then. Maybe we could move this script to the rh-syft repo in the future? I don't think it belongs in Cachi2.

chmeliik replied:
> Maybe we could move this script to the rh-syft repo in the future?

The fewer things we productize (and have to support) the better. I wanted the `syft` executable to be the only thing in the image, but I may end up including a ubi-micro base. But probably nothing past that.

If we do need a script long-term, it should probably be tightly coupled to the task, versioned together with the task etc. Though it shouldn't be inlined in the yaml and it should be tested.

chmeliik replied:
Btw, one of the cataloggers that `syft` includes is `sbom`. Maybe we don't need a script for syft X syft deduplication at all

brunoapimentel replied:
I did some early testing and the sobm cataloger seems promising. I will file a story to investigate it further.

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,237 @@
+    name: IMAGE_DIGEST
+  - description: Image repository where the built image was pushed
+    name: IMAGE_URL
+  - description: Digests of the base images used for build
+    name: BASE_IMAGES_DIGESTS
```
As far as I could see, this is only used in the [deprecated-base-image-check](https://github.com/redhat-appstudio/build-definitions/blob/main/pipelines/template-build/template-build.yaml#L137-L147), which is [removed](https://github.com/redhat-appstudio/build-definitions/blob/main/pipelines/docker-build-dance/patch.yaml#L67-L68) from the RHTAP pipeline.

Any chance this is used in any other place?

chmeliik replied:
Yeah, it's used for an EC check https://enterprisecontract.dev/docs/ec-policies/release_policy.html#base_image_registries_package

brunoapimentel replied:
So I think we'll need to keep this for now, and probably move STONEBLD-2042 into the SBOM epic.

mmorhun commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,197 @@
+      with open("./sbom-cyclonedx.json", "rb") as f:
+        sbom_digest = hashlib.file_digest(f, "sha256").hexdigest()
+
+      image_repo = os.getenv("IMAGE").split(":")[0]
```
It will fail if registry has port, e.g. `registry.com:1234/user/image:tag`

brunoapimentel replied:
I dropped this commit, let's handle the SBOM_BLOB_URL in a follow up PR.

mmorhun commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,197 @@
+            component_list.append(component)
+            existing_purls.append(component["purl"])
+        else:
+          # We won't try to dedup components that lack a purl.
```
```suggestion
          # We won't try to deduplicate components that lack a purl.
```

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,197 @@
+        exit 1
+      fi
+
+      sed -i 's/^\s*short-name-mode\s*=\s*.*/short-name-mode = "disabled"/' /etc/containers/registries.conf
```
I copied this from the current buildah task without giving it much thought. Thing is, apparently this enables short-name resolution (see related [commit](https://github.com/redhat-appstudio/build-definitions/commit/018608552f790f4d1bd529f366173be13ee7bd51)). 

I don't think we want to keep this for the RHTAP task, as it is less secure than requiring full names. Wdyt?

chmeliik replied:
Oh, good catch, I thought this line was there to _disable_ short name search :exploding_head: 

I think we need "enforcing" https://man.archlinux.org/man/containers-registries.conf.5.en#Short-Name_Aliasing:_Modes, which is the default in the redhat buildah image

```
$ podman run --rm -ti registry.access.redhat.com/ubi9/buildah@sha256:04fde77ea72c25b56efb3f71db809c5d7b09938130df2da9175a3c888b94043d cat /etc/containers/registries.conf

...
short-name-mode = "enforcing"
```

brunoapimentel replied:
Since the deafult is enforcing, I removed this line here.

chmeliik commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,197 @@
+      sbom_blob_url = f"{image_repo}@sha256:{sbom_digest}"
+
+      print(image_repo)
+      print(sbom_blob_url)
```
Are the prints intentional?

brunoapimentel replied:
Nope, they were accidentally commited in some rebase :face_exhaling:

brunoapimentel replied:
This commit was entirely dropped.

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,197 @@
+    image: quay.io/redhat-appstudio/syft:v0.98.0@sha256:4d3856e6a2622700b9a9d5d74d9aaf5d8a55671653f80bf6c636677658680ede
+    script: |
+      syft dir:$(workspaces.source.path)/source --output cyclonedx-json@1.4=/tmp/files/sbom-source.json
+      syft oci-dir:/tmp/files/image --output cyclonedx-json@1.4=/tmp/files/sbom-image.json
```
I already changed this locally to 1.5, will push the fix later.

brunoapimentel replied:
Fixed.

mmorhun reviewed (approved):
I think we can start from current state of the new Buildah task and fix found problems in follow up PRs.

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,197 @@
+    securityContext:
+      capabilities:
+        add:
+          - SETFCAP
```
Added.

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,197 @@
+
+      for component in source_sbom.get("components", []):
+        if "purl" in component:
+          if component["purl"] not in existing_purls:
```
You're absolutely right, I stopped halfway through that filter thing. It should be right now (this part at least... deduplication will need major rework)

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -68,25 +68,23 @@ spec:
-      find $(cat /workspace/container_path) -xtype l -delete
-
-      syft dir:$(cat /workspace/container_path) --output cyclonedx-json=/tmp/files/sbom-image.json
+      syft oci-dir:/tmp/sboms/image --output cyclonedx-json=/tmp/files/sbom-image.json
```
I caught one instance of a `/tmp/sboms` leftover and fixed it. Not sure if I overlooked more commit-specific errors.

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,197 @@
+      buildah push \
+        --tls-verify=$TLSVERIFY \
+        --digestfile /tmp/files/image-digest $IMAGE \
+        docker://$IMAGE
```
I added the --retry option, and kept it to 5 as it was the number set before.

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,197 @@
+      value: $(params.TLSVERIFY)
+  steps:
+  - image: registry.access.redhat.com/ubi9/buildah@sha256:04fde77ea72c25b56efb3f71db809c5d7b09938130df2da9175a3c888b94043d
+    name: build
```
Fixed.

brunoapimentel commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,197 @@
+    name: build
+    script: |
+      # Check if the Dockerfile exists
+      SOURCE_CODE_DIR=source
```
I will add this to the list of follow up stories.

brunoapimentel commented:
The newer push is just rebasing the branch. We still need to create the repository for the new task's image, since the CI is failing because of it.

mmorhun commented:
/retest build-definitions-pull-request

chmeliik commented:
/retest

chmeliik commented:
@arewm it looks like your change request is blocking the merge button

arewm commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,174 @@
+      name: tmpfiles
+    workingDir: $(workspaces.source.path)
+
+  - name: generate-sboms
+    image: quay.io/redhat-appstudio/syft:v0.98.0@sha256:4d3856e6a2622700b9a9d5d74d9aaf5d8a55671653f80bf6c636677658680ede
```
@brunoapimentel , is modifying this a follow up story as well?

brunoapimentel replied:
Do you mean changing this for the productized version Adam is working on? If so, yes, it should be a follow up.

arewm replied:
yes, great!

arewm commented on task/buildah-rhtap/0.1/buildah-rhtap.yaml:
```diff
@@ -0,0 +1,174 @@
+      name: tmpfiles
+    workingDir: /tmp/files
+
+  - name: upload-sbom
+    image: quay.io/redhat-appstudio/cosign:v2.1.1@sha256:c883d6f8d39148f2cea71bff4622d196d89df3e510f36c140c097b932f0dd5d5
```
I think we should be able to modify this to the RHTAS images

brunoapimentel replied:
Do you have a link for them?

arewm replied:
https://catalog.redhat.com/software/containers/rhtas-tech-preview/cosign-rhel9/652971cd584d77e9d922883b?architecture=amd64&image=65c5506e19b6942d637e5603
